chalk-solve/src/solve/slg/implementation/clauses.rs

```diff
@@ -0,0 +1,47 @@
+use chalk_ir::*;
+use chalk_ir::cast::{Cast, Caster};
+
+crate fn program_clauses_for_dynamic_types(goal: &DomainGoal) -> impl Iterator<Item = ProgramClause> {
```

Comment:
Interesting. I was wondering how to handle the "open-ended" set of rules here. 

I was imagining perhaps a slightly more "meta" approach, but it would require deeper changes to the engine. 

e.g., I was imagining something where to solve `dyn (\Self => C): Trait<..>` we kind of convert it to:

```rust
forall<Self> {
  if (C[Self]) { // here C is assumed to be `Self => Clause`
    Self: Trait<..>
  }
}
```

Reply:
In rustc we already have this approach where we return program clauses only when we have a bit more information about the goal, because of e.g. tuples and fn types, for which the arity cannot be encoded in the logic.

Reply:
Yep. I think I probably like this approach better than baking dyn types more deeply into the engine.

Reply:
It feels to me like -- at some point -- we should definitely try to have more code sharing between chalk + rustc though. There is a fair amount of non-trivial logic that seems like it is basically the same between the two projects.